docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,242 @@
+    MetricName string `json:"metricName"`
+    // the current value of the metric (as a quantity)
+    CurrentValue resource.Quantity `json:"targetValue"`
+}
+
+// metric describing pods in the scale target
+type PodsMetricStatus struct {
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the current value of the metric (as a quantity)
+    CurrentValue resource.Quantity `json:"targetValue"`
+}
+```
+
+### Example ###
+
+In this example, we scale based on the `hits-per-second` value recorded as
+describing a service in our namespace, plus the CPU usage of the pods in
+the ReplicationController being autoscaled.
+
+```yaml
+kind: HorizontalPodAutoscaler
+apiVersion: autoscaling/v2alpha1
+spec:
+  scaleTargetRef:
+    kind: ReplicationController
+    name: WebFrontend
+  minReplicas: 2
+  maxReplicas: 10
+  targetCPUUtilizationPercentage: 80
+  metrics:
+  - object:
+      target:
+        kind: Service
+        name: Frontend
+      metricName: hits-per-second
+      targetValue: 1k
+```
+
+### Alternatives and Future Considerations ###
+
+Since the new design mirrors volume plugins (and similar APIs), it makes
+it relatively easy to introduce new fields in a backwards-compatible way:
+we simply introduce a new field in `MetricSpec` as a new "metric type".
+
+#### External ####
+
+As a potential future consideration, we may wish to include a metric
+source type of `External`, which simply has an opaque metric type and
+target value (as strings).  This would allow custom autoscalers with
+custom source mechanics to be written, while still providing the same
+indication that a controller was being autoscaled.
+
+#### Resource Percentages ####
+
+In the current design, CPU percentage metrics are a completely separate
+field from the rest of the metrics.  While this simplifies the common case
+of scaling on CPU, it makes the overall design a bit more clunky.  It also
+doesn't provide a method for scaling based on memory percentage.
+
+Two alternatives to the method proposed above are shown the object below.
+The former is more compact, but precludes scaling memory as a percentage
+of request.
+
+```yaml
+kind: HorizontalPodAutoscaler
+apiVersion: autoscaling/v2alpha1
+spec:
+  scaleTargetRef:
+    kind: ReplicationController
+    name: WebFrontend
+  minReplicas: 2
+  maxReplicas: 10
+  metrics:
+  - cpuRequest:
+      targetPercentage: 80
+  - resourceRequest:
+      resource: cpu
+      targetPercentage: 80
+```
+
+#### Referring to the current Namespace ####
+
+It is beneficial to be able to refer to a metric on the current namespace,
+similarly to the `ObjectMetricSource` source type, but without an explicit
+name.  Because of the similarity to `ObjectMetricSource`, it may simply be
+sufficient to allow specificying a `kind` of "Namespace" without a name.
+Alternatively, a similar source type to `PodsMetricSource` could be used.
+
+#### Aggregating Desired Replica Counts ####
```

Comment:
I think it would extremely confusing for user to reason about the actions of HPA if we add custom aggregators between metrics.